inox2d/src/node/drawables.rs

```diff
@@ -0,0 +1,66 @@
+use glam::Mat4;
+
+use crate::node::{
+	components::{Composite, Drawable, TexturedMesh, TransformStore},
+	InoxNodeUuid,
+};
+use crate::puppet::World;
+
+/// Possible component combinations of a renderable node.
+///
+/// Future spec extensions go here.
+/// For user-defined custom nodes that can be rendered, as long as a subset of their components matches one of these variants,
+/// they will be picked up and enter the regular rendering pipeline.
+pub(crate) enum DrawableKind<'comps> {
+	TexturedMesh(TexturedMeshComponents<'comps>),
+	Composite(CompositeComponents<'comps>),
+}
+
+/// Pack of components for a TexturedMesh. "Part" in Inochi2D terms.
+pub struct TexturedMeshComponents<'comps> {
+	// Only the absolute part of `TransformStore` that the renderer backend may need.
+	pub transform: &'comps Mat4,
+	pub drawable: &'comps Drawable,
+	pub data: &'comps TexturedMesh,
+}
+
+/// Pack of components for a Composite node.
+pub struct CompositeComponents<'comps> {
+	// Only the absolute part of `TransformStore` that the renderer backend may need.
+	pub transform: &'comps Mat4,
+	pub drawable: &'comps Drawable,
+	pub data: &'comps Composite,
+}
+
+impl<'comps> DrawableKind<'comps> {
+	/// Tries to construct a renderable node data pack from the World of components:
+	/// - `None` if node not renderable.
+	/// - Panicks if component combinations invalid.
+	pub(crate) fn new(id: InoxNodeUuid, comps: &'comps World) -> Option<Self> {
+		let drawable = match comps.get::<Drawable>(id) {
+			Some(drawable) => drawable,
+			None => return None,
+		};
+		let transform = &comps
+			.get::<TransformStore>(id)
+			.expect("A drawble must have an associated transform.")
+			.absolute;
+		let textured_mesh = comps.get::<TexturedMesh>(id);
+		let composite = comps.get::<Composite>(id);
+
+		match (textured_mesh.is_some(), composite.is_some()) {
+			(true, true) => panic!("The drawable has both TexturedMesh and Composite."),
+			(false, false) => panic!("The drawable has neither TexturedMesh nor Composite."),
```

Comment:
A node having both a `TexturedMesh` and a `Composite` is something that is up to the model, so it's something that can actually happen. Crashing doesn't make sense here.
I propose that we return something that seems wrong instead of panicking. For example `(false, true) | (true, true) => Some(DrawableKind::Composite(...))` (which gives priority to the composite and ignores the textured mesh) and `(false, false) => None`.
Then somewhere else during initialization, we can log an error with tracing to warn the user of a malformed of corrupted model, and even go as far as to say that the textured mesh will be ignored.

Reply:
A node having both components is out of Inochi2D spec. It is us who choose to represent `class Part : Drawable` and `class Composite : Drawable` in the D ref impl with composition in Rust. If it is type wise not possible to describe a node with both (you can't override a same function twice!) in D, I guess it is not proper for Inox2D core to do so.

If you are persuaded that this is out of spec, and still wants to keep such behavior as Inox2D extension, then my opinion on such an extension is: Keep having both of these components a signal of a bug on our end or an improper usage from the end user. Use components of other types for representing such special nodes that act as both textured mesh and composite.

Reply:
It being out of the spec got past me, but it is not really a good reason to add points of panic. inox2d is only responsible for moving puppets, it's weird and unexpected for such a library to crash the end user's program. In our case it's better to continue with a wrong-ish state because the end user can still run their application without it crashing, which would be more annoying. Think textures not loading being replaced by black-magenta default textures instead of blowing up and not running. So I maintain that if it has a problem it should ideally log it and move on as best it can.